src/compiler/checker.ts

```diff
@@ -21689,7 +21689,27 @@ namespace ts {
             if (type.flags & TypeFlags.Union) {
                 const types = (<UnionType>type).types;
                 const filtered = filter(types, f);
-                return filtered === types ? type : getUnionTypeFromSortedList(filtered, (<UnionType>type).objectFlags);
+                if (filtered === types) {
+                    return type;
+                }
+                const origin = (<UnionType>type).origin;
+                let newOrigin: Type | undefined;
+                if (origin && origin.flags & TypeFlags.Union) {
+                    // If the origin type is a (denormalized) union type, filter its non-union constituents. If that ends
+                    // up removing a smaller number of types than in the normalized constituent set (meaning some of the

```

Comment:
This says smaller, but you're checking for whether or not you've reduced the same number of elements.

Reply:
It's either going to be smaller or the same. It could never be larger.

Reply:
I guess what I'm saying is that the code is written in an inverted way, so it wasn't as obvious what's happening. I just realized it was written this way to share the same `return` path though.

Reply:
`createOriginUnionOrIntersectionType` is the thing you *can't* use if you end up removing a smaller number of types, right? If the length diff is the same, then you can, otherwise you have to fall back to `getUnionTypeFromSortedList`, which will ignore the origin information.

Reply:
Oh, github didn't load the previous 3 comments before I commented. I guess the point is that I think the comment is not confusing, even if the code itself is.